### ISSUE TITLE
Allow parallel execution of functions

### ISSUE THREAD

**Issue by rdmchr** [appwrite/appwrite#1425]
## 🚀 Feature

I believe that it would be nice if one function could run multiple times at the same time.

## Have you spent some time to check if this issue has been raised before?

Yes.

### Have you read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/master/CODE_OF_CONDUCT.md)?

Yes.

## Pitch

Let's say I'm running a function that does some OCR on all uploaded images. One run of the function usually takes 25 seconds. In my use case, users will be able to scan some handwritten stuff and the text will be saved into a database. Since I expect users to submit images in batches of let's say 10. This means that a user will have to wait over 4 minutes for their papers to get scanned.
Since my server can handle 4 parallel executions this could be reduced to about 1 minute.

I'm not sure about the implementation but from the looks of it you would definitely need to change the naming scheme of function containers.

**Comment by eldadfux** [comment]
@rdmchr you can get more parallel executions by adding more functions worker containers in your docker stack. For example:

```yaml
...
appwrite-worker-functions-1:
    entrypoint: worker-functions
    container_name: appwrite-worker-functions
    build:
      context: .
    networks:
      - appwrite
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - appwrite-functions:/storage/functions:rw
      - /tmp:/tmp:rw
      - ./app:/usr/src/code/app
      - ./src:/usr/src/code/src
    depends_on:
      - redis
      - mariadb
    environment:
      - _APP_ENV
      - _APP_OPENSSL_KEY_V1
      - _APP_REDIS_HOST
      - _APP_REDIS_PORT
      - _APP_REDIS_USER
      - _APP_REDIS_PASS
      - _APP_DB_HOST
      - _APP_DB_PORT
      - _APP_DB_SCHEMA
      - _APP_DB_USER
      - _APP_DB_PASS
      - _APP_FUNCTIONS_TIMEOUT
      - _APP_FUNCTIONS_CONTAINERS
      - _APP_FUNCTIONS_RUNTIMES
      - _APP_FUNCTIONS_CPUS
      - _APP_FUNCTIONS_MEMORY
      - _APP_FUNCTIONS_MEMORY_SWAP
      - _APP_USAGE_STATS
      - DOCKERHUB_PULL_USERNAME
      - DOCKERHUB_PULL_PASSWORD
 
appwrite-worker-functions-2:
    entrypoint: worker-functions
    container_name: appwrite-worker-functions
    build:
      context: .
    networks:
      - appwrite
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - appwrite-functions:/storage/functions:rw
      - /tmp:/tmp:rw
      - ./app:/usr/src/code/app
      - ./src:/usr/src/code/src
    depends_on:
      - redis
      - mariadb
    environment:
      - _APP_ENV
      - _APP_OPENSSL_KEY_V1
      - _APP_REDIS_HOST
      - _APP_REDIS_PORT
      - _APP_REDIS_USER
      - _APP_REDIS_PASS
      - _APP_DB_HOST
      - _APP_DB_PORT
      - _APP_DB_SCHEMA
      - _APP_DB_USER
      - _APP_DB_PASS
      - _APP_FUNCTIONS_TIMEOUT
      - _APP_FUNCTIONS_CONTAINERS
      - _APP_FUNCTIONS_RUNTIMES
      - _APP_FUNCTIONS_CPUS
      - _APP_FUNCTIONS_MEMORY
      - _APP_FUNCTIONS_MEMORY_SWAP
      - _APP_USAGE_STATS
      - DOCKERHUB_PULL_USERNAME
      - DOCKERHUB_PULL_PASSWORD
 ...
```

The upcoming update to functions will introduce sync execution where you could get the execution response immediately. Using this model, scaling will behave more similarly to the API. It will use multiple co-routines for parallel executions.

**Comment by rdmchr** [comment]
Hmm, I see. Ill definitely make sure to try that out if it makes sense for my use case. I figured that the way you described might work, but Meldiron and I figured that docker would not allow you to start two containers with the same name and it seems like the name is relatively [hardcoded](https://github.com/appwrite/appwrite/blob/28f0c5f39fd97e5f8d1dabb781482b74e27c239c/app/workers/functions.php#L372).
But I'm hyped for the functions update ;)

**Comment by TorstenDittmann** [comment]
We just released 0.13 and we completely re-wrote the Functions Service where this is possible out-of-the-box now 👍🏻 

You can find the announcement [here](https://dev.to/appwrite/announcing-appwrite-013-with-major-upgrades-to-storage-and-functions-3hpf).

Thank you a lot for reporting this 🙏🏻

**Comment by stnguyen90** [comment]
@TorstenDittmann, can this be re-opened? I don't think this is possible for executions triggered via events. It would be nice to have an option on the function for whether it should run in parallel or sequentially.

**Comment by TorstenDittmann** [comment]
> @TorstenDittmann, can this be re-opened? I don't think this is possible for executions triggered via events. It would be nice to have an option on the function for whether it should run in parallel or sequentially.

I am currently looking into replacing resque, the new implementation will allow us to do parallel executions across all workers 👍🏻

**Comment by voodee** [comment]
> @rdmchr you can get more parallel executions by adding more functions worker containers in your docker stack. For example:
> 
> ```yaml
> ...
> appwrite-worker-functions-1:
>     entrypoint: worker-functions
>     container_name: appwrite-worker-functions
>     build:
>       context: .
>     networks:
>       - appwrite
>     volumes:
>       - /var/run/docker.sock:/var/run/docker.sock
>       - appwrite-functions:/storage/functions:rw
>       - /tmp:/tmp:rw
>       - ./app:/usr/src/code/app
>       - ./src:/usr/src/code/src
>     depends_on:
>       - redis
>       - mariadb
>     environment:
>       - _APP_ENV
>       - _APP_OPENSSL_KEY_V1
>       - _APP_REDIS_HOST
>       - _APP_REDIS_PORT
>       - _APP_REDIS_USER
>       - _APP_REDIS_PASS
>       - _APP_DB_HOST
>       - _APP_DB_PORT
>       - _APP_DB_SCHEMA
>       - _APP_DB_USER
>       - _APP_DB_PASS
>       - _APP_FUNCTIONS_TIMEOUT
>       - _APP_FUNCTIONS_CONTAINERS
>       - _APP_FUNCTIONS_RUNTIMES
>       - _APP_FUNCTIONS_CPUS
>       - _APP_FUNCTIONS_MEMORY
>       - _APP_FUNCTIONS_MEMORY_SWAP
>       - _APP_USAGE_STATS
>       - DOCKERHUB_PULL_USERNAME
>       - DOCKERHUB_PULL_PASSWORD
>  
> appwrite-worker-functions-2:
>     entrypoint: worker-functions
>     container_name: appwrite-worker-functions
>     build:
>       context: .
>     networks:
>       - appwrite
>     volumes:
>       - /var/run/docker.sock:/var/run/docker.sock
>       - appwrite-functions:/storage/functions:rw
>       - /tmp:/tmp:rw
>       - ./app:/usr/src/code/app
>       - ./src:/usr/src/code/src
>     depends_on:
>       - redis
>       - mariadb
>     environment:
>       - _APP_ENV
>       - _APP_OPENSSL_KEY_V1
>       - _APP_REDIS_HOST
>       - _APP_REDIS_PORT
>       - _APP_REDIS_USER
>       - _APP_REDIS_PASS
>       - _APP_DB_HOST
>       - _APP_DB_PORT
>       - _APP_DB_SCHEMA
>       - _APP_DB_USER
>       - _APP_DB_PASS
>       - _APP_FUNCTIONS_TIMEOUT
>       - _APP_FUNCTIONS_CONTAINERS
>       - _APP_FUNCTIONS_RUNTIMES
>       - _APP_FUNCTIONS_CPUS
>       - _APP_FUNCTIONS_MEMORY
>       - _APP_FUNCTIONS_MEMORY_SWAP
>       - _APP_USAGE_STATS
>       - DOCKERHUB_PULL_USERNAME
>       - DOCKERHUB_PULL_PASSWORD
>  ...
> ```
> 
> The upcoming update to functions will introduce sync execution where you could get the execution response immediately. Using this model, scaling will behave more similarly to the API. It will use multiple co-routines for parallel executions.

I tried this method and added 10 containers. Most of the time it works, but sometimes an error appears:
```
Docker Error: docker: Error response from daemon: Conflict. The container name "/6254191e8556e5a51efe-62651942426226a145d7" is already in use by container "d3c8b63d3a5f578f24efcda12dac79d8caa6203b74293926e3dc2b65db98a4f2". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
```

**Comment by voodee** [comment]
Maybe this error is also related to this problem.

![Appwrite erro](https://user-images.githubusercontent.com/831470/165986081-09327cb1-4037-4ebc-ab31-c94ab50012ae.png)

This error appeared after adding the containers.

`docker-compose restart` solves this problem.

**Comment by PineappleIOnic** [comment]
Parallel executions are supported natively in 0.13 since we use multi-threaded web servers for most runtimes. 

@voodee There is no longer any need to increase the number of function workers in Appwrite, this can actually cause issues due to the internal runtime state. The issue you are facing should be fixed in 0.15 thanks to #3361

@stnguyen90 0.14 introduced the new event model which should have fixed your issue, any other problems don't be afraid to create a new issue 🙂

**Comment by voodee** [comment]
@PineappleIOnic it's not working for me 😞

If I make one copy of "appwrite-worker-functions", only one task is performed.
If I make two copies of "appwrite-worker-functions", then two tasks run in parallel.

![Appwrite - Function (4)](https://user-images.githubusercontent.com/831470/184197728-946d8af0-3511-46b8-b8d7-634faab0a948.png)

**Comment by voodee** [comment]
@PineappleIOnic open this issue, pls. This feature is really important for me.

**Comment by PineappleIOnic** [comment]
@voodee What runtime are you using?

**Comment by troykelly** [comment]
For clarity - should we be seeing multiple function execution currently, or is there an outstanding bug?
From my reading of the thread - multiple executors were a solution, but that now breaks things - so it can no longer be used.

Is my reading of the above that there is no way to have multiple executors and self-hosted AppWrite can/will only execute one function at a time?

**Comment by fliitor** [comment]
> > @TorstenDittmann, can this be re-opened? I don't think this is possible for executions triggered via events. It would be nice to have an option on the function for whether it should run in parallel or sequentially.
> 
> I am currently looking into replacing resque, the new implementation will allow us to do parallel executions across all workers 👍🏻

Any news ?

**Comment by lisu-wefind-ai** [comment]
Hi, I still can not find any update about this.

And even worse,  I found the execution from multiple Functions can only execute one by one?  It means I can only run one execution (other new execution will be in waiting status) at the same time even the executions come from many diffrent Functions.

What's wrong with my settings? It's really unacceptable!

**Comment by DHsoftw** [comment]
@lisu-wefind-ai 
Set async to true when executing the function: https://appwrite.io/docs/client/functions?sdk=web-default#functionsCreateExecution

**Comment by lisu-wefind-ai** [comment]
> @lisu-wefind-ai Set async to true when executing the function: https://appwrite.io/docs/client/functions?sdk=web-default#functionsCreateExecution

Thanks bro.  What I find is just like the things bellow.
I use it good for my purpose, run multiple executions the same time from the web console.
And notion: I changed the code's container_name to make it work.

> @rdmchr you can get more parallel executions by adding more functions worker containers in your docker stack. For example:
> 
> ```yaml
> ...
> appwrite-worker-functions-1:
>     entrypoint: worker-functions
>     container_name: appwrite-worker-functions-1
/
>     build:
>       context: .
>     networks:
>       - appwrite
>     volumes:
>       - /var/run/docker.sock:/var/run/docker.sock
>       - appwrite-functions:/storage/functions:rw
>       - /tmp:/tmp:rw
>       - ./app:/usr/src/code/app
>       - ./src:/usr/src/code/src
>     depends_on:
>       - redis
>       - mariadb
>     environment:
>       - _APP_ENV
>       - _APP_OPENSSL_KEY_V1
>       - _APP_REDIS_HOST
>       - _APP_REDIS_PORT
>       - _APP_REDIS_USER
>       - _APP_REDIS_PASS
>       - _APP_DB_HOST
>       - _APP_DB_PORT
>       - _APP_DB_SCHEMA
>       - _APP_DB_USER
>       - _APP_DB_PASS
>       - _APP_FUNCTIONS_TIMEOUT
>       - _APP_FUNCTIONS_CONTAINERS
>       - _APP_FUNCTIONS_RUNTIMES
>       - _APP_FUNCTIONS_CPUS
>       - _APP_FUNCTIONS_MEMORY
>       - _APP_FUNCTIONS_MEMORY_SWAP
>       - _APP_USAGE_STATS
>       - DOCKERHUB_PULL_USERNAME
>       - DOCKERHUB_PULL_PASSWORD
>  
> appwrite-worker-functions-2:
>     entrypoint: worker-functions
>     container_name: appwrite-worker-functions-2
>     build:
>       context: .
>     networks:
>       - appwrite
>     volumes:
>       - /var/run/docker.sock:/var/run/docker.sock
>       - appwrite-functions:/storage/functions:rw
>       - /tmp:/tmp:rw
>       - ./app:/usr/src/code/app
>       - ./src:/usr/src/code/src
>     depends_on:
>       - redis
>       - mariadb
>     environment:
>       - _APP_ENV
>       - _APP_OPENSSL_KEY_V1
>       - _APP_REDIS_HOST
>       - _APP_REDIS_PORT
>       - _APP_REDIS_USER
>       - _APP_REDIS_PASS
>       - _APP_DB_HOST
>       - _APP_DB_PORT
>       - _APP_DB_SCHEMA
>       - _APP_DB_USER
>       - _APP_DB_PASS
>       - _APP_FUNCTIONS_TIMEOUT
>       - _APP_FUNCTIONS_CONTAINERS
>       - _APP_FUNCTIONS_RUNTIMES
>       - _APP_FUNCTIONS_CPUS
>       - _APP_FUNCTIONS_MEMORY
>       - _APP_FUNCTIONS_MEMORY_SWAP
>       - _APP_USAGE_STATS
>       - DOCKERHUB_PULL_USERNAME
>       - DOCKERHUB_PULL_PASSWORD
>  ...
> ```
> 
> The upcoming update to functions will introduce sync execution where you could get the execution response immediately. Using this model, scaling will behave more similarly to the API. It will use multiple co-routines for parallel executions.

**Comment by EVDOG4LIFE** [comment]
Closing out due to age as this has been implemented, if this is still valid I recommend creating a new issue as we are quite a few versions ahead now.
